keras/layers/layer.py

```diff
@@ -945,11 +947,14 @@ def stateless_call(
             )
 
         # Gather variable mapping
-        trainable_mapping = zip(self.trainable_variables, trainable_variables)
-        non_trainable_mapping = zip(
-            self.non_trainable_variables, non_trainable_variables
+        all_variables = map(
+            lambda v: v.value if isinstance(v, KerasVariable) else v,
```

Comment:
For greater generality we should do this in `StatelessScope` I think?

What I don't get is that we already do `v = backend.convert_to_tensor(v, dtype=k.dtype)` in `StatelessScope`. This should grab the value, if I'm not mistaken?

Reply:
> For greater generality we should do this in `StatelessScope` I think?

Yes, I don't like the current duplication. And it's totally trivial to add.

> What I don't get is that we already do `v = backend.convert_to_tensor(v, dtype=k.dtype)` in `StatelessScope`. This should grab the value, if I'm not mistaken?

Oh, now I fully understand why my change to `tensorflow.core.convert_to_tensor` triggered this.

The issue is that in Tensorflow, we have:
```python
class Variable(
    KerasVariable,
    tf.__internal__.types.Tensor,
    tf.__internal__.tracking.Trackable,
):
```
Because of the `tf.__internal__.types.Tensor`, `tf.is_tensor` returns True and bypasses the conversion. However, we were lucky enough that `tf.cast` doesn't have a shortcut when the dtypes are identical, so it would cause the cast to happen always, at which point the variable was turned to an actual `tf.Tensor`.

Reply:
But isn't the issue with JAX specifically? I'm confused.

Reply:
Ok, perhaps not fully related, but I did run into exactly this issue with JAX a bit earlier -- you cannot call `model.stateless_call` with Variables, you need to unwrap them. The nature of the issue is an infinite recursion, which seems to happen at the level of JAX.

Can you just add explicit `if isinstance` logic in `StatelessScope`? That should fix it.

Reply:
Yes, doing that, now it's a 1-line change basically. Thanks!